src/visitors/local_var_visitor.cpp

```diff
@@ -0,0 +1,84 @@
+/*************************************************************************
+ * Copyright (C) 2018-2019 Blue Brain Project
+ *
+ * This file is part of NMODL distributed under the terms of the GNU
+ * Lesser General Public License. See top-level LICENSE file for details.
+ *************************************************************************/
+
+#include <iostream>
+#include <memory>
+#include <unordered_set>
+
+#include "ast/assigned_block.hpp"
+#include "ast/assigned_definition.hpp"
+#include "ast/local_list_statement.hpp"
+#include "ast/local_var.hpp"
+#include "ast/node.hpp"
+#include "ast/program.hpp"
+#include "visitors/local_var_visitor.hpp"
+#include "visitors/visitor_utils.hpp"
+
+namespace nmodl {
+namespace visitor {
+
+void LocalToAssignedVisitor::visit_program(ast::Program& node) {
+    ast::AssignedDefinitionVector assigned_variables;
+    std::unordered_set<ast::LocalVar*> local_variables_to_remove;
+    std::unordered_set<ast::Node*> local_nodes_to_remove;
+    std::shared_ptr<ast::AssignedBlock> assigned_block;
+
+    const auto& top_level_nodes = node.get_blocks();
+    const auto& symbol_table = node.get_symbol_table();
+
+    for (auto& top_level_node: top_level_nodes) {
+        /// only process local_list statements
+        if (top_level_node->is_local_list_statement()) {
+            const auto& local_variables =
+                std::static_pointer_cast<ast::LocalListStatement>(top_level_node)->get_variables();
+            for (auto& local_variable: local_variables) {
```

Comment:
const auto& could be clearer

Reply:
I think that we had a similar conversation in the `GlobalToRange` PR. @pramodk asked to remove the unnecessary `const` declaration and since some functions return `const ref`s I should maybe remove a couple of `const` from the code

Reply:
You're right.. one just has to remember that these functions actually return const refs, that's a bit the annoying part. I have no clear preference but wonder what people typically do...

Reply:
>I have no clear preference but wonder what people typically do...

I had exactly the same question but I couldn't find any clear answer on the webz. I will keep looking though

Reply:
I still didn't find any solid answer on the internet. It seems that there is absolutely no functional difference between the two. It's just a matter of preference.
To me it looks more readable with the `const` it front as well. Maybe this is a policy we should or shouldn't follow? 
I would also like to hear @tristan0x and @pramodk opinion on this

Reply:
Given a C++ class exposing a non-modifiable member variable through a member function, if its API evolves and now needs to allow the member variable to be modified, then **it should provide an additional member function** rather than removing the `const` specifier in the first one.

For instance:

```c++
class LocalListStatement {
  const LocalVarVector& get_variables() const;
  LocalVarVector& get_variables();
};
```

Reply:
In this case, I am much more bothered by the instruction below than the `const-ness` of the loop.

```
const auto& local_variables =
                std::static_pointer_cast<ast::LocalListStatement>(top_level_node)->get_variables();
```

1. The very same call to `std::static_pointer_case` is made 10 lines below, so the result shared_ptr should be stored in a local variable.
1. It lacks clarity

Reply:
If the code was manipulating a  `std::static_pointer_cast<const ast::LocalListStatement>`, then IMO I would not be so strict at repeating `const` everywhere afterward.

Buts since the result of `std::static_pointer_cast` returns a shared_ptr of a modifiable `ast::LocalListStatement`, IMO code manipulating it should indicate whether it is modifying it or not.

So I agree with @cattabiani that the range-based loop should indicate it is not modifying `local_variable`.

Reply:
According to my comments, here is what I would change:

```diff
diff --git a/src/visitors/local_var_visitor.cpp b/src/visitors/local_var_visitor.cpp
index ef4dd74..f2b2d42 100644
--- a/src/visitors/local_var_visitor.cpp
+++ b/src/visitors/local_var_visitor.cpp
@@ -30,12 +30,11 @@ void LocalToAssignedVisitor::visit_program(ast::Program& node) {
     auto& top_level_nodes = node.get_blocks();
     const auto& symbol_table = node.get_symbol_table();

-    for (auto& top_level_node: top_level_nodes) {
+    for (const auto& top_level_node: top_level_nodes) {
         /// only process local_list statements
         if (top_level_node->is_local_list_statement()) {
-            auto& local_variables =
-                std::static_pointer_cast<ast::LocalListStatement>(top_level_node)->get_variables();
-            for (auto& local_variable: local_variables) {
+            const auto& node = std::static_pointer_cast<ast::LocalListStatement>(top_level_node);
+            for (const auto& local_variable: node->get_variables()) {
                 auto variable_name = local_variable->get_node_name();
                 /// check if local variable is being updated in the mod file
                 if (symbol_table->lookup(variable_name)->get_write_count() > 0) {
@@ -52,11 +51,10 @@ void LocalToAssignedVisitor::visit_program(ast::Program& node) {
             }

             /// remove offending local variables
-            std::static_pointer_cast<ast::LocalListStatement>(top_level_node)
-                ->erase_local_var(local_variables_to_remove);
+            node->erase_local_var(local_variables_to_remove);

             /// add empty local statements to local_statements_to_remove
-            if (local_variables.empty()) {
+            if (node->get_variables().empty()) {
                 local_nodes_to_remove.emplace(top_level_node.get());
             }
         }
```

Reply:
I see.. I agree with @tristan0x points and with that `const` should be added in the `for-loop` iterators. Tristan's suggestion looks nice and clean to me, so I am going to adopt it. Thank you Tristan! 🙂

Reply:
Just out of curiosity, what is about using forms like:

```cpp
     const auto& top_level_nodes = node.get_blocks();
     for (auto& top_level_node: top_level_nodes) {
```
i.e. not using `const` in for loop iterator as your var is already const.